### PR TITLE
Add field to schmea

### PR DESCRIPTION
**Pull request by bbrewer-papa-x**
schmeeeeeeea

### CODE REVIEW

File: apps/andi/config/integration.exs [hunk 1]
```diff
@@ -67,3 +76,20 @@ config :andi, AndiWeb.Endpoint,
   live_view: [
     signing_salt: "SUPER VERY TOP SECRET!!!"
   ]
+
+defmodule Andi.DivoPostgres do
```
Comment (bbrewer-papa-x) [inline apps/andi/config/integration.exs:80]: we can replace this with https://hexdocs.pm/divo_postgres/readme.html#content before we're wrapped up

File: apps/andi/lib/andi/event_handler.ex [hunk 2]
```diff
@@ -15,7 +15,15 @@ defmodule Andi.EventHandler do
   @ingested_time_topic "ingested_time_topic"
 
   def handle_event(%Brook.Event{type: dataset_update(), data: %Dataset{} = data}) do
-    DatasetCache.put(data)
+    dataset =
+      case Andi.Repo.get(Andi.InputSchemas.Dataset, data.id) do
```
Comment (bbrewer-papa-x) [inline apps/andi/lib/andi/event_handler.ex:19]: we'll be putting this under the repo pattern eventually. Something like `Andi.Schemas.Datasets.get_or_create`

File: apps/andi/lib/andi/input_schemas/dataset.ex [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+    field(:systemName, :string)
+    field(:sourceUrl, :string)
+    field(:topLevelSelector, :string)
+    field(:cadence, :string)
+    field(:orgId, :string)
+
+    belongs_to(:dataset, Andi.InputSchemas.Dataset, type: :string, primary_key: true, foreign_key: :dataset_id)
+  end
+
+  def changeset(technical, %SmartCity.Dataset.Technical{} = changes) do
```
Comment (bbrewer-papa-x) [inline apps/andi/lib/andi/input_schemas/dataset.ex:27]: we'll break these out and also move the conversion related stuff back into InputConverter

File: apps/andi/lib/andi/input_schemas/dataset.ex [hunk 4]
```diff
@@ -0,0 +1,292 @@
+defmodule Andi.InputSchemas.Technical do
```
Comment (LtChae) [inline apps/andi/lib/andi/input_schemas/dataset.ex:1]: This file feels massive. Haven't done much Ecto, but is there a way to break this up?

File: apps/andi/assets/css/app.scss [hunk 5]
```diff
@@ -142,3 +142,6 @@ body {
 
 @import "./datasets.scss";
 @import "./edit.scss";
+@import "./data_dictionary_form.scss";
+@import "./metadata_form.scss";
+@import "./url_form.scss";
```
Comment (cameronmarsh) [inline apps/andi/assets/css/app.scss:147]: Broke these out into their own filees

File: apps/andi/config/integration.exs [hunk 6]
```diff
@@ -12,11 +12,20 @@ endpoint = [{host, 9092}]
...
+    {DivoRedis, []},
+    {Andi.DivoPostgres, []}
   ],
   divo_wait: [dwell: 700, max_tries: 50],
   kafka_broker: endpoint
 
+config :andi, Andi.Repo,
+  database: "andi",
+  username: "postgres",
+  password: "postgres",
```
Comment (dunson062786) [inline apps/andi/config/integration.exs:24]: Probably want to change the two values above and make them environment variables.

File: apps/andi/lib/andi/input_schemas/data_dictionary_fields.ex [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+      _ ->
+        field
+        |> Map.put(:bread_crumb, parent_bread_crumb <> " > " <> field.name)
+    end
+  end
+
+  def get_parent_ids(dataset) do
+    dataset_id = dataset.id
+    top_cheddar = [{@top_level_bread_crumb, dataset.technical.id}]
```
Comment (LtChae) [inline apps/andi/lib/andi/input_schemas/data_dictionary_fields.ex:42]: Paul would be proud

File: apps/andi/lib/andi/input_schemas/datasets/query_param.ex [hunk 8]
```diff
@@ -0,0 +1,39 @@
+defmodule Andi.InputSchemas.Datasets.QueryParam do
```
Comment (LtChae) [inline apps/andi/lib/andi/input_schemas/datasets/query_param.ex:1]: This seems to be identical to Headers. Could Technical have two lists of the same type of child?

File: apps/andi/lib/andi/input_schemas/input_converter.ex [hunk 9]
```diff
@@ -3,84 +3,180 @@ defmodule Andi.InputSchemas.InputConverter do
   Used to convert between SmartCity.Datasets, form data (defined by Andi.InputSchemas.DatasetInput), and Ecto.Changesets.
   """
 
-  alias SmartCity.Dataset
-  alias Andi.InputSchemas.DatasetInput
+  alias Andi.InputSchemas.Datasets.Dataset
```
Comment (LtChae) [inline apps/andi/lib/andi/input_schemas/input_converter.ex:6]: Should the converter methods live with the schemas they convert from/to?

File: apps/andi/assets/css/metadata_form.scss [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  color: $color-error;
+}
+
+.metadata__success-message {
+  text-align: right;
+  padding-right: 2em;
+  color: $color-success;
+}
+
+// @include area(metadata-form, top-header);
```
Comment (JarredOlson) [inline apps/andi/assets/css/metadata_form.scss:28]:
```suggestion
```

File: apps/andi/test/unit/andi_web/live/edit_live_view_test.exs [hunk 11]
```diff
@@ -149,126 +162,143 @@ defmodule AndiWeb.EditLiveViewTest do
...
-      assert get_value(html, "#form_data_license") == dataset.business.license
-      assert get_value(html, "#form_data_publishFrequency") == dataset.business.publishFrequency
-      assert get_value(html, "#form_data_spatial") == dataset.business.spatial
-      assert get_value(html, "#form_data_temporal") == dataset.business.temporal
-      assert get_value(html, "#form_data_orgTitle") == dataset.business.orgTitle
-      assert {"english", "English"} == get_select(html, "#form_data_language")
-      assert get_value(html, "#form_data_homepage") == dataset.business.homepage
-      assert {"1.0", "High"} == get_select(html, "#form_data_benefitRating")
-      assert {"0.5", "Medium"} == get_select(html, "#form_data_riskRating")
+      assert get_value(html, ".metadata-form__title input") == dataset.business.dataTitle
```
Comment (dunson062786) [inline apps/andi/test/unit/andi_web/live/edit_live_view_test.exs:255]: would it be cleaner to store these selectors in an external file and import them when you need them.

File: apps/andi/lib/andi/event_handler.ex [hunk 12]
```diff
@@ -9,14 +9,15 @@ defmodule Andi.EventHandler do
...
 
+  alias Andi.InputSchemas.Datasets
+
   @ingested_time_topic "ingested_time_topic"
 
   def handle_event(%Brook.Event{type: dataset_update(), data: %Dataset{} = data}) do
-    DatasetCache.put(data)
-    {:merge, :dataset, data.id, data}
+    Datasets.update(data)
+    DatasetStore.update(data)
```
Comment (JarredOlson) [inline apps/andi/lib/andi/event_handler.ex:20]: Why do we seemingly have two different storages of a dataset?

File: apps/andi/lib/andi/input_schemas/data_dictionary_fields.ex [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+    dd_query =
```
Comment (JarredOlson) [inline apps/andi/lib/andi/input_schemas/data_dictionary_fields.ex:44]: please change `dd` to `dataset_dictionary` or `data_dictionary`

File: apps/andi/lib/andi/input_schemas/data_dictionary_fields.ex (hunk 7 above)
Comment (JarredOlson) [reply apps/andi/lib/andi/input_schemas/data_dictionary_fields.ex:42]: Agreed.  Please change

File: apps/andi/config/integration.exs (hunk 6 above)
Comment (cameronmarsh) [reply apps/andi/config/integration.exs:24]: These are just config values so we can spin up ecto in the integration environment. This is how we are doing it in discovery_api

Comment (kcarra) [reply apps/andi/config/integration.exs:24]: this is just used for integration tests. No need to worry about security

File: apps/andi/test/unit/andi_web/live/edit_live_view_test.exs (hunk 11 above)
Comment (cameronmarsh) [reply apps/andi/test/unit/andi_web/live/edit_live_view_test.exs:255]: Eh I don't know, I think they are helpful in the assertions to tell which piece of the view we are looking at

File: apps/andi/lib/andi/input_schemas/datasets/query_param.ex (hunk 8 above)
Comment (cameronmarsh) [reply apps/andi/lib/andi/input_schemas/datasets/query_param.ex:1]: I think the idea is that we can store them separately in ecto so we can query against them without complicating the queries and get calls. But you're right, there's a lot of duplication in here that I'll try to clean up

File: apps/andi/lib/andi/input_schemas/input_converter.ex (hunk 9 above)
Comment (cameronmarsh) [reply apps/andi/lib/andi/input_schemas/input_converter.ex:6]: I'm not sure these functions are specific enough to one of the input schemas to do that. We could probably move some of them, but it made sense to consolidate them to one place, imo. But I can leave that up to discussion

File: apps/andi/lib/andi/input_schemas/data_dictionary_fields.ex [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+  def add_field_to_parent(original_field, parent_bread_crumb) do
+    updated_field = adjust_parent_details(original_field, parent_bread_crumb)
+    changeset = DataDictionary.changeset(%DataDictionary{}, updated_field)
+
+    case Repo.insert_or_update(changeset) do
+      {:error, _changeset} ->
+        {:error, DataDictionary.changeset(%DataDictionary{}, original_field)}
+
+      good ->
```
Comment (kcarra) [inline apps/andi/lib/andi/input_schemas/data_dictionary_fields.ex:21]: I'm assuming this wasn't intentional?

File: apps/andi/lib/andi/input_schemas/struct_tools.ex [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+    value
+  end
+
+  def preload(nil, _fields), do: nil
+
+  def preload(list, fields) when is_list(list) do
+    Enum.map(list, &preload(&1, fields))
+  end
+
+  def preload(%struct_type{} = struct, fields) do
```
Comment (kcarra) [inline apps/andi/lib/andi/input_schemas/struct_tools.ex:36]: :eyes: I want to know whats going on here, I really do, but I don't

File: apps/andi/lib/andi/input_schemas/data_dictionary_fields.ex (hunk 14 above)
Comment (bbrewer-papa-x) [reply apps/andi/lib/andi/input_schemas/data_dictionary_fields.ex:21]: That was me being me

File: apps/andi/lib/andi/input_schemas/struct_tools.ex (hunk 15 above)
Comment (cameronmarsh) [reply apps/andi/lib/andi/input_schemas/struct_tools.ex:36]: This is basically loading all of the associations of a struct in ecto based on the type of each of those associations. It's pretty gross but necessary. I can work on cleaning this up a little

**Pull request merged**